dotnet/src/MicroFocus.CafApi.QueryBuilder.Matcher/MatcherQueryVisitor.cs

```diff
@@ -145,83 +241,228 @@ public void VisitBetween(IMatcherFieldSpec<Document> fieldSpec, string startValu
+                string valueToCompare = fieldSpec.IsCaseInsensitive ? value.ToUpperInvariant() : value;
 
-            _isMatch = GetStringValues(fieldSpec).Any(v => v.CompareTo(valueToCompare) >= 0);
+                _isMatch = values.Any(v => v.CompareTo(valueToCompare) >= 0);
+            }
         }
 
         public void VisitExists(IMatcherFieldSpec<Document> fieldSpec)
         {
-            _isMatch = GetFieldValues(fieldSpec).Any();
+            var values = GetFieldValues(fieldSpec);
+            if (values == null)
+            {
+                _isMatch = false;
+            }
+            else
+            {
+                _isMatch = values.Any();
+            }
         }
 
         public void VisitEmpty(IMatcherFieldSpec<Document> fieldSpec)
         {
-            _isMatch = !GetStringValues(fieldSpec).Any(v => v.Length > 0);
+            var values = GetStringValues(fieldSpec);
+            if (values == null)
+            {
+                _isMatch = null;
+            }
+            else
+            {
+                _isMatch = !values.Any(v => v.Length > 0);
+            }
         }
 
         public void VisitOr(IEnumerable<Filter<IMatcherFieldSpec<Document>>> filters)
         {
+            /*
+             true or true === true
+             true or false === true
+             false or true === true
+             false or false === false
+
+             true or unknown === true
+             false or unknown === unknown
+             unknown or true === true
+             unknown or false === unknown
+
+             unknown or unknown === unknown
+            */
+            var matches = new List<bool?>();
             foreach (Filter<IMatcherFieldSpec<Document>> filter in filters)
             {
                 filter.Invoke(this);
-                if (_isMatch)
-                {
-                    break;
-                }
+                matches.Add(_isMatch);
+            }
+            if (matches.Contains(true))
+            {
+                _isMatch = true;
+            }
+            else if (matches.Contains(null))
+            {
+                _isMatch = null;
+            }
+            else
+            {
+                _isMatch = false;
             }
         }
 
         public void VisitAnd(IEnumerable<Filter<IMatcherFieldSpec<Document>>> filters)
         {
-            _isMatch = filters.All(filter => filter.IsMatch(_document, _allFullTextFieldSpecs));
+            /*
+             true and true === true
+             true and false === false
+             false and true === false
+             false and false === false
+
+             true and unknown === unknown
+             false and unknown === false
+             unknown and true === unknown
+             unknown and false === false
+
+             unknown and unknown === unknown
+           */
+
+            var matches = new List<bool?>();
+            foreach (Filter<IMatcherFieldSpec<Document>> filter in filters)
```

Comment:
We're losing short-circuiting here too.

Reply:
public void VisitAnd(IEnumerable<Filter<IMatcherFieldSpec<Document>>> filters)
    {
        bool? result = true;

        foreach (Filter<IMatcherFieldSpec<Document>> filter in filters)
        {
            bool? isMatch = filter.IsMatch(_document, _allFullTextFieldSpecs);

            if (isMatch == false)
            {
                result = false;
                break;
            }

            if (isMatch == null)
            {
                result = null;
            }
        }

        _isMatch = result;
    }